flambe/nn/transformer.py

```diff
@@ -0,0 +1,532 @@
+        memory_key_padding_mask: torch.Tensor, optional
+            the ByteTensor mask for memory keys per batch (optional).
+            shape" :math:`(N, S)`.
+
+        Returns
+        -------
+        output: torch.Tensor
+            The output sequence, shape: :math:`(T, N, E)`.
+
+        Note: [src/tgt/memory]_mask should be filled with
+            float('-inf') for the masked positions and float(0.0) else.
+            These masks ensure that predictions for position i depend
+            only on the unmasked positions j and are applied identically
+            for each sequence in a batch.
+            [src/tgt/memory]_key_padding_mask should be a ByteTensor
+            where True values are positions that should be masked with
+            float('-inf') and False values will be unchanged.
+            This mask ensures that no information will be taken from
+            position i if it is masked, and has a separate mask for each
+            sequence in a batch.
+        Note: Due to the multi-head attention architecture in the
+            transformer model, the output sequence length of a
+            transformer is same as the input sequence
+            (i.e. target) length of the decode.
+
+            where S is the source sequence length, T is the target
+            sequence length, N is the batchsize, E is the feature number
+
+        """
+
+        if src.size(1) != tgt.size(1):
+            raise RuntimeError("the batch number of src and tgt must be equal")
+
+        if src.size(2) != self.d_model or tgt.size(2) != self.d_model:
+            raise RuntimeError("the feature number of src and tgt must be equal to d_model")
+
+        memory = self.encoder(src,
+                              mask=src_mask,
+                              src_key_padding_mask=src_key_padding_mask)
+        output = self.decoder(tgt,
+                              memory,
+                              tgt_mask=tgt_mask,
+                              memory_mask=memory_mask,
+                              tgt_key_padding_mask=tgt_key_padding_mask,
+                              memory_key_padding_mask=memory_key_padding_mask)
+        return output
+
+    def generate_square_subsequent_mask(self, sz):
+        r"""Generate a square mask for the sequence.
+
+        The masked positions are filled with float('-inf').
+        Unmasked positions are filled with float(0.0).
+        """
+        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
+        mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+        return mask
+
+
+class TransformerEncoder(Module):
+    """TransformerEncoder is a stack of N encoder layers."""
+
+    def __init__(self,
+                 d_model: int,
+                 nhead: int,
+                 num_layers: int,
+                 dim_feedforward: int = 2048,
+                 dropout: float = 0.1,
+                 use_sru: bool = False) -> None:
```

Comment:
A different option is to provide `custom_forward_encoder: Module = None` as a flexible choice to replace FFN.
For example, depending on the application, the SRUCell can be uni-directional or bi-directional. Or we may want to explore a different hyper-parameter setup for SRUCell. This avoids adding SRU args to the __init__ method?

You can create  `nn.TransformerSRU` which then creates SRUCell() for each layer and pass that as the custom_forward_encoder?

Reply:
TransformerSRU may also support you to pass the hidden states of the last data segment (chunk) as the init hidden states of the current data segment. 

This is useful for training corpus language model with BPTT.

Reply:
Yeah so the question here is that if I make a TransformerSRU, I also need to make a TransformerSRUEncoder, TransformerSRUDecoder, etc.. Seems like a lot of repeated code thats hard to avoid.

I considered the idea of passing a custom_forward_module. We could definitely do that, or take additional **kwargs to be passed to SRU. I think the custom_forward_module idea makes the most sense so that you can pass in anything you want, doesn't even have to be SRU. Thoughts?

Reply:
I like the idea of custom_forward_module and very simple TransformerSRU.

Regarding the state, maybe i should add that to the main implementation, and the feedforward doesn't use it?

Reply:
> Regarding the state, maybe i should add that to the main implementation, and the feedforward doesn't use it?
you mean add it as an optional value and default to None? sounds good to me.